### PR TITLE
enhance(monitor):add druid monitor

### PR DESCRIPTION
**Pull request by ungreat**
#### What type of PR is this?
add durid spring method and sql stat monitor, and alarm  while the num exceeds the threshold,

#### Which issue(s) this PR fixes:

Fixes #

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-server/src/main/resources/log4j2.xml [hunk 1]
```diff
@@ -15,11 +15,15 @@
              如果每个类的 log 对象都基于当前类生成，%logger 可作为 %class 的替代。
         -->
         <Property name="LOG_PATTERN">
-            [%date{yyyy-MM-dd HH:mm:ss.SSS}][%thread][%X{traceId},%X{spanId},%X{requestId}][%level][%class][%line]: %message%n
+            [%date{yyyy-MM-dd HH:mm:ss.SSS}][%thread][%X{traceId},%X{spanId},%X{requestId}][%level][%class][%line]:
```
Comment (yizhouxw) [inline server/odc-server/src/main/resources/log4j2.xml:18]: here seems only a line breaker change, let's check the xml code format configuration

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+import com.oceanbase.odc.core.alarm.AlarmUtils;
+import com.oceanbase.odc.service.config.SystemConfigService;
+import com.oceanbase.odc.service.config.model.Configuration;
+
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+
+@Component
+@Slf4j
+@ConditionalOnProperty(value = "odc.monitor.enabled", havingValue = "true")
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:58]: may use key `odc.system.monitor.enabled` .
odc may offer user side datatabase monitor features in future.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+            vo.setJdbcFetchRowCount((Long) map.get(DruidMethodStatConstants.JDBC_FETCH_ROW_COUNT));
+            vo.setJdbcUpdateCount((Long) map.get(DruidMethodStatConstants.JDBC_UPDATE_COUNT));
+            vo.setLastError(map.get(DruidMethodStatConstants.LAST_ERROR));
+            vo.setLastErrorTime((Date) map.get(DruidMethodStatConstants.LAST_ERROR_TIME));
+            vo.setHistogram((long[]) map.get(DruidMethodStatConstants.HISTOGRAM));
+            return vo;
+        }
+    }
+
+    public static String compressSql(String sql) {
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:620]: please add some unit testing cases.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 4]
```diff
@@ -0,0 +1,646 @@
...
+        }
+
+        public String buildTooMuchJdbcExecuteCountMsg() {
+            return getMethodIdentification() + ", avgJdbcExecuteCount=" + computeAvgJdbcExecuteCount()
+                    + ";";
+        }
+
+        public static DruidMethodStatVO mapToVO(Map<String, Object> map) {
+            DruidMethodStatVO vo = new DruidMethodStatVO();
+            vo.setClazz((String) map.get(DruidMethodStatConstants.CLASS));
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:593]: will `map.get()` return null ?

**Review by yizhouxw — COMMENTED** [review]
great job 👍

File: server/odc-service/src/main/java/com/oceanbase/odc/config/DruidConfig.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import org.springframework.context.annotation.Configuration;
+
+import com.alibaba.druid.support.spring.stat.DruidStatInterceptor;
+
+@Configuration
+public class DruidConfig {
+
+    @Bean
+    public DruidStatInterceptor druidStatInterceptor() {
+        DruidStatInterceptor dsInterceptor = new DruidStatInterceptor();
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/config/DruidConfig.java:29]: move this to `com.oceanbase.odc.config.MetaDBConfiguration`?

File: server/odc-server/src/main/resources/log4j2.xml (hunk 1 above)
Comment (yhilmare) [inline server/odc-server/src/main/resources/log4j2.xml:18]: why change this?

File: server/odc-service/src/main/java/com/oceanbase/odc/config/druid/MethodInvocationAdapter.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+package com.oceanbase.odc.config.druid;
+
+import java.lang.reflect.AccessibleObject;
+import java.lang.reflect.Method;
+
+import org.aopalliance.intercept.MethodInvocation;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.reflect.MethodSignature;
+
+public class MethodInvocationAdapter implements MethodInvocation {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/config/druid/MethodInvocationAdapter.java:26]: what does this interceptor do? I can not get it by its name

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 7]
```diff
@@ -0,0 +1,646 @@
...
+
+    @Autowired
+    private SystemConfigService systemConfigService;
+
+    private List<Configuration> configurations;
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        druidStatManagerFacade.setResetEnable(true);
+        Executors.newScheduledThreadPool(1).scheduleAtFixedRate(() -> {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:83]: when should this executor be shutdown?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 8]
```diff
@@ -0,0 +1,646 @@
...
+        }, 1, 2, TimeUnit.MINUTES);
+    }
+
+    public void doMonitor() {
+        configurations = systemConfigService.queryByKeyPrefix(CONFIG_PREFIX);
+        monitorMethod();
+        monitorSqlStatData();
+        monitorDruidDataSource();
+    }
+
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:99]: empty lines

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 9]
```diff
@@ -0,0 +1,646 @@
...
+        private final String key;
+
+        DruidMonitorConfig(String key, String defaultValue) {
+            Preconditions.checkArgument(key.startsWith(CONFIG_PREFIX));
+            this.key = key;
+            this.defaultValue = defaultValue;
+        }
+    }
+
+    public static final class DruidSQLStateConstants {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:206]: is 'public' flag necessary ?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 10]
```diff
@@ -0,0 +1,646 @@
...
+        public static final String EXECUTE_COUNT = "ExecuteCount";
+        public static final String EFFECTED_ROW_COUNT = "EffectedRowCount";
+        public static final String READ_STRING_LENGTH = "ReadStringLength";
+        public static final String EXECUTE_AND_RESULT_HOLD_TIME_HISTOGRAM = "ExecuteAndResultHoldTimeHistogram";
+        public static final String FILE = "File";
+        public static final String CLOB_OPEN_COUNT = "ClobOpenCount";
+        public static final String LAST_TIME = "LastTime";
+        public static final String FETCH_ROW_COUNT_HISTOGRAM = "FetchRowCountHistogram";
+
+        private DruidSQLStateConstants() {}
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:245]: why should we declare it?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 11]
```diff
@@ -0,0 +1,646 @@
...
+    }
+
+    @Data
+    public static class DruidSQLStateVO {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:249]: VO,DTO... such a name is not recommended in odc, how about remove 'VO' suffix?(DruidSqlState)

and, is 'public' necessary?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 12]
```diff
@@ -0,0 +1,646 @@
...
+            vo.setBlobOpenCount((Long) map.get(DruidSQLStateConstants.BLOB_OPEN_COUNT));
+            vo.setExecuteCount((Long) map.get(DruidSQLStateConstants.EXECUTE_COUNT));
+            vo.setEffectedRowCount((Long) map.get(DruidSQLStateConstants.EFFECTED_ROW_COUNT));
+            vo.setReadStringLength((Long) map.get(DruidSQLStateConstants.READ_STRING_LENGTH));
+            vo.setExecuteAndResultHoldTimeHistogram(
+                    (long[]) map.get(DruidSQLStateConstants.EXECUTE_AND_RESULT_HOLD_TIME_HISTOGRAM));
+            vo.setFile((String) map.get(DruidSQLStateConstants.FILE));
+            vo.setClobOpenCount((Long) map.get(DruidSQLStateConstants.CLOB_OPEN_COUNT));
+            vo.setLastTime((Date) map.get(DruidSQLStateConstants.LAST_TIME));
+            vo.setFetchRowCountHistogram((long[]) map.get(DruidSQLStateConstants.FETCH_ROW_COUNT_HISTOGRAM));
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:416]: how to guarantee that the value type of the map(Map<String, Object>)?

how should we know that the type of value(Map<String, Object>) has been changed if we upgrade the druid's version?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 13]
```diff
@@ -0,0 +1,646 @@
...
+        monitorMethodJdbcExecuteTooLong(methodStatVOS);
+        monitorMethodTooMuchJdbcExecute(methodStatVOS);
+
+    }
+
+    private void monitorMethodAvgExecuteTooLong(List<DruidMethodStatVO> methodStatVOS) {
+        List<String> msg = methodStatVOS.stream()
+                .filter(t -> t.computeAvgExecuteTime() > getLongMonitorConfig(DruidMonitorConfig.METHOD_EXECUTE_TIME))
+                .map(
+                        DruidMethodStatVO::buildTooLongAvgExecuteTimeMsg)
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:443]: should the alert message be international？

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 14]
```diff
@@ -0,0 +1,646 @@
...
+                        DruidMethodStatVO::buildTooMuchJdbcExecuteCountMsg)
+                .collect(
+                        Collectors.toList());
+        if (!msg.isEmpty()) {
+            AlarmUtils.alarm(METHOD_TOO_MUCH_JDBC_EXECUTE_COUNT, "method too much jdbc execute count" + msg);
+        }
+    }
+
+    @Data
+    public static final class DruidMethodStatConstants {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:477]: is ‘public’ necessary?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 15]
```diff
@@ -0,0 +1,646 @@
...
+        public static final String JDBC_EXECUTE_COUNT = "JdbcExecuteCount";
+        public static final String JDBC_EXECUTE_ERROR_COUNT = "JdbcExecuteErrorCount";
+        public static final String JDBC_EXECUTE_TIME_MILLIS = "JdbcExecuteTimeMillis";
+        public static final String JDBC_FETCH_ROW_COUNT = "JdbcFetchRowCount";
+        public static final String JDBC_UPDATE_COUNT = "JdbcUpdateCount";
+        public static final String LAST_ERROR = "LastError";
+        public static final String LAST_ERROR_TIME = "LastErrorTime";
+        public static final String HISTOGRAM = "Histogram";
+
+        private DruidMethodStatConstants() {}
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:502]: why should we declare this constructor?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 16]
```diff
@@ -0,0 +1,646 @@
...
+    }
+
+    @Data
+    public static class DruidMethodStatVO {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:506]: VO is not recommended

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 17]
```diff
@@ -0,0 +1,646 @@
...
+            return getMethodIdentification() + ", avgJdbcExecuteTime=" + computeAvgJdbcExecuteTime() + ";";
+        }
+
+        public String buildTooMuchJdbcExecuteCountMsg() {
+            return getMethodIdentification() + ", avgJdbcExecuteCount=" + computeAvgJdbcExecuteCount()
+                    + ";";
+        }
+
+        public static DruidMethodStatVO mapToVO(Map<String, Object> map) {
+            DruidMethodStatVO vo = new DruidMethodStatVO();
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:592]: same as above

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java [hunk 18]
```diff
@@ -0,0 +1,646 @@
...
+            vo.setHistogram((long[]) map.get(DruidMethodStatConstants.HISTOGRAM));
+            return vo;
+        }
+    }
+
+    public static String compressSql(String sql) {
+        if (sql == null) {
+            return null;
+        }
+        final int maxLength = 100;
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:624]: what does these magic number mean?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java (hunk 3 above)
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:620]: we may add a method to do this, how about 'StringUtils#getBriefSql'?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java (hunk 15 above)
Comment (ungreat) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:502]: Utility classes, which are collections of static members, are not meant to be instantiated.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java (hunk 12 above)
Comment (ungreat) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:416]: Indeed, it is not possible to automatically ensure the correct correspondence of Map<String, Object> , so it needs to be manually judged whether the data has correct value by observing the logs such as method-monitor.log or sql-monitor.log. Based on the current statistical methods of Druid, there is no better approach for the time being.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java (hunk 10 above)
Comment (ungreat) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:245]: Actually, it's a rule of SonarLint，which says `Utility classes should not have public constructors`

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java (hunk 2 above)
Comment (ungreat) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java (hunk 4 above)
Comment (ungreat) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java:619]: Some fields should semantically be null, whereas other fields related to statistical values have already been given default values.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java (hunk 7 above)
Comment (ungreat) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java]: add preDestroy hook to gracefulShutdown

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java (hunk 3 above)
Comment (ungreat) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/config/druid/MethodInvocationAdapter.java (hunk 6 above)
Comment (ungreat) [reply server/odc-service/src/main/java/com/oceanbase/odc/config/druid/MethodInvocationAdapter.java]: Clear for JointPointMethodInvocationAdapter.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java (hunk 3 above)
Comment (ungreat) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/common/DruidMonitor.java]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/config/druid/MethodInvocationAdapter.java (hunk 6 above)
Comment (yhilmare) [reply server/odc-service/src/main/java/com/oceanbase/odc/config/druid/MethodInvocationAdapter.java]: It seems like that this methodInterceptor does nothing

**Review by yhilmare — APPROVED** [review]
lgtm

**Pull request merged**